File: semaphore.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <cstdio>

using namespace std;

class Semaphore {
public:
    Semaphore(int initialCount) : count(initialCount) {}

    void acquire() {
        std::unique_lock<std::mutex> lock(mutex);
        while (count == 0) {
            condition.wait(lock);
        }
        count--;
    }

    void release() {
        std::lock_guard<std::mutex> lock(mutex);
        count++;
        condition.notify_one();
    }

private:
    int count;
    std::mutex mutex;
    std::condition_variable condition;
};

Semaphore semaphore(2); // Initialize semaphore with a count of 2

void worker(int id) {
    semaphore.acquire();
    std::cout << "Worker " << id << " acquired the semaphore." << std::endl;
    std::this_thread::sleep_for(std::chrono::seconds(2)); // Simulate some work
    std::cout << "Worker " << id << " releasing the semaphore." << std::endl;
    semaphore.release();
}

int main() {

    std::thread t1(worker, 1);
    std::thread t2(worker, 2);
    std::thread t3(worker, 3);

    t1.join();
    t2.join();
    t3.join();

    return 0;
}

File: semaphore/semaphore.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type Semaphore struct {
	count int
	mutex sync.Mutex
	cond  *sync.Cond
}

func NewSemaphore(initialCount int) *Semaphore {
	s := &Semaphore{count: initialCount}
	s.cond = sync.NewCond(&s.mutex)
	return s
}

func (s *Semaphore) Acquire() {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	for s.count == 0 {
		s.cond.Wait()
	}
	s.count--
}

func (s *Semaphore) Release() {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	s.count++
	s.cond.Signal()
}

var semaphore = NewSemaphore(2)

func worker(id int) {
	semaphore.Acquire()
	fmt.Printf("Worker %d acquired the semaphore.\n", id)
	time.Sleep(time.Second * 5) // Simulate some work
	fmt.Printf("Worker %d releasing the semaphore.\n", id)
	semaphore.Release()
}

func main() {
	var wg sync.WaitGroup

	wg.Add(3)
	go func() {
		defer wg.Done()
		worker(1)
	}()

	go func() {
		defer wg.Done()
		worker(2)
	}()

	go func() {
		defer wg.Done()
		worker(3)
	}()

	wg.Wait()
}
